### OBJECTIVE: Identify frequent itemsets using Apriori Algorithm.

Apriori algorithm is a machine learning model used in Association Rule Learning to identify frequent itemsets from a dataset. This model has been highly applied on transactions datasets by large retailers to determine items that customers frequently buy together with high probability.

The Apriori algorithm uses three matrices to find the best association rules from a dataset:
Support - It measures the number of times a particular item or combination of items occur in a dataset. 
Confidence - It measures how the consumer is likely to consume commodity x given that they have consumed commodity y. 
Lift - A lift is a metric that determines the strength of association between the best rules. It is obtained by taking confidence and diving it with support.

In [1]:
#Importing necessary libraries
from mlxtend.frequent_patterns import apriori,association_rules
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
import time

In [2]:
#Importing and reading the dataset
data = [['citrus fruit', 'semi-finished bread', 'margarine', 'ready soups'],
       ['tropical fruit', 'yogurt', 'coffee'], 
       ['whole milk'], 
       ['pip fruit', 'yogurt', 'cream cheese', 'meat spreads'], 
       ['other vegetables', 'whole milk', 'condensed milk', 'long life bakery product'],
       ['whole milk', 'butter', 'yogurt', 'rice', 'abrasive cleaner'],
       ['rolls/buns'], 
       ['other vegetables', 'UHT-milk', 'rolls/buns', 'bottled beer', 'liquor (appetizer)'],
       ['potted plants'], 
       ['whole milk', 'cereals'], 
       ['tropical fruit', 'other vegetables', 'white bread', 'bottled water', 'chocolate'],
       ['citrus fruit', 'tropical fruit', 'whole milk', 'butter', 'curd', 'yogurt', 'flour', 'bottled water', 'dishes'],
       ['beef'], ['frankfurter', 'rolls/buns', 'soda'], 
       ['chicken', 'tropical fruit'], 
       ['butter', 'sugar', 'fruit/vegetable juice', 'newspapers'], 
       ['fruit/vegetable juice'],
       ['packaged fruit/vegetables'], 
       ['chocolate'], 
       ['specialty bar'], 
       ['other vegetables'], 
       ['butter milk', 'pastry'],
       ['whole milk'], 
       ['tropical fruit', 'cream cheese', 'processed cheese', 'detergent', 'newspapers']]
data

[['citrus fruit', 'semi-finished bread', 'margarine', 'ready soups'],
 ['tropical fruit', 'yogurt', 'coffee'],
 ['whole milk'],
 ['pip fruit', 'yogurt', 'cream cheese', 'meat spreads'],
 ['other vegetables',
  'whole milk',
  'condensed milk',
  'long life bakery product'],
 ['whole milk', 'butter', 'yogurt', 'rice', 'abrasive cleaner'],
 ['rolls/buns'],
 ['other vegetables',
  'UHT-milk',
  'rolls/buns',
  'bottled beer',
  'liquor (appetizer)'],
 ['potted plants'],
 ['whole milk', 'cereals'],
 ['tropical fruit',
  'other vegetables',
  'white bread',
  'bottled water',
  'chocolate'],
 ['citrus fruit',
  'tropical fruit',
  'whole milk',
  'butter',
  'curd',
  'yogurt',
  'flour',
  'bottled water',
  'dishes'],
 ['beef'],
 ['frankfurter', 'rolls/buns', 'soda'],
 ['chicken', 'tropical fruit'],
 ['butter', 'sugar', 'fruit/vegetable juice', 'newspapers'],
 ['fruit/vegetable juice'],
 ['packaged fruit/vegetables'],
 ['chocolate'],
 ['specialty bar'],
 ['other vegetables'],
 ['butter mi

In [3]:
#Converting the dataset into pandas dataframe
data1 = pd.DataFrame(data)
data1

0                    1                      2  \
0                citrus fruit  semi-finished bread              margarine   
1              tropical fruit               yogurt                 coffee   
2                  whole milk                 None                   None   
3                   pip fruit               yogurt           cream cheese   
4            other vegetables           whole milk         condensed milk   
5                  whole milk               butter                 yogurt   
6                  rolls/buns                 None                   None   
7            other vegetables             UHT-milk             rolls/buns   
8               potted plants                 None                   None   
9                  whole milk              cereals                   None   
10             tropical fruit     other vegetables            white bread   
11               citrus fruit       tropical fruit             whole milk   
12                       beef                 None                   None   
13                frankfurter           rolls/buns                   soda   
14                    chicken       tropical fruit                   None   
15                     butter                sugar  fruit/vegetable juice   
16      fruit/vegetable juice                 None                   None   
17  packaged fruit/vegetables                 None                   None   
18                  chocolate                 None                   None   
19              specialty bar                 None                   None   
20           other vegetables                 None                   None   
21                butter milk               pastry                   None   
22                 whole milk                 None                   None   
23             tropical fruit         cream cheese       processed cheese   

                           3                   4       5      6  \
0                ready soups                None    None   None   
1                       None                None    None   None   
2                       None                None    None   None   
3               meat spreads                None    None   None   
4   long life bakery product                None    None   None   
5                       rice    abrasive cleaner    None   None   
6                       None                None    None   None   
7               bottled beer  liquor (appetizer)    None   None   
8                       None                None    None   None   
9                       None                None    None   None   
10             bottled water           chocolate    None   None   
11                    butter                curd  yogurt  flour   
12                      None                None    None   None   
13                      None                None    None   None   
14                      None                None    None   None   
15                newspapers                None    None   None   
16                      None                None    None   None   
17                      None                None    None   None   
18                      None                None    None   None   
19                      None                None    None   None   
20                      None                None    None   None   
21                      None                None    None   None   
22                      None                None    None   None   
23                 detergent          newspapers    None   None   

                7       8  
0            None    None  
1            None    None  
2            None    None  
3            None    None  
4            None    None  
5            None    None  
6            None    None  
7            None    None  
8            None    None  
9            None    None  
10           None    None  
11  bottled water  dishes  
12           None    None  
13           None    None  
14    

In [5]:
#Finding the most popular items that has been purchased
popular=data1[0].value_counts()
popular

whole milk                   4
tropical fruit               3
other vegetables             3
citrus fruit                 2
pip fruit                    1
rolls/buns                   1
potted plants                1
beef                         1
frankfurter                  1
chicken                      1
butter                       1
fruit/vegetable juice        1
packaged fruit/vegetables    1
chocolate                    1
specialty bar                1
butter milk                  1
Name: 0, dtype: int64

In [6]:
#Initializing the transactionEncoder
te = TransactionEncoder()
te_ary = te.fit(data).transform(data)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

UHT-milk  abrasive cleaner   beef  bottled beer  bottled water  butter  \
0      False             False  False         False          False   False   
1      False             False  False         False          False   False   
2      False             False  False         False          False   False   
3      False             False  False         False          False   False   
4      False             False  False         False          False   False   
5      False              True  False         False          False    True   
6      False             False  False         False          False   False   
7       True             False  False          True          False   False   
8      False             False  False         False          False   False   
9      False             False  False         False          False   False   
10     False             False  False         False           True   False   
11     False             False  False         False           True    True   
12     False             False   True         False          False   False   
13     False             False  False         False          False   False   
14     False             False  False         False          False   False   
15     False             False  False         False          False    True   
16     False             False  False         False          False   False   
17     False             False  False         False          False   False   
18     False             False  False         False          False   False   
19     False             False  False         False          False   False   
20     False             False  False         False          False   False   
21     False             False  False         False          False   False   
22     False             False  False         False          False   False   
23     False             False  False         False          False   False   

    butter milk  cereals  chicken  chocolate  ...   rice  rolls/buns  \
0         False    False    False      False  ...  False       False   
1         False    False    False      False  ...  False       False   
2         False    False    False      False  ...  False       False   
3         False    False    False      False  ...  False       False   
4         False    False    False      False  ...  False       False   
5         False    False    False      False  ...   True       False   
6         False    False    False      False  ...  False        True   
7         False    False    False      False  ...  False        True   
8         False    False    False      False  ...  False       False   
9         False     True    False      False  ...  False       False   
10        False    False    False       True  ...  False       False   
11        False    False    False      False  ...  False       False   
12        False    False    False      False  ...  False       False   
13        False    False    False      False  ...  False        True   
14        False    False     True      False  ...  False       False   
15        False    False    False      False  ...  False       False   
16        False    False    False      False  ...  False       False   
17        False    False    False      False  ...  False       False   
18        False    False    False       True  ...  False       False   
19        False    False    False      False  ...  False       False   
20        False    False    False      False  ...  False       False   
21         True    False    False      False  ...  False       False   
22        False    False    False      False  ...  False       False   
23        False    False    False      False  ...  False       False   

    semi-finished bread   soda  specialty bar  sugar  tropical fruit  \
0                  True  False          False  False           False   
1                 False  False          False  False            True   
2                 False  False      

In [7]:
#Picking frequent itemsets
apriori(df, min_support=0.05)

support     itemsets
0   0.083333          (4)
1   0.125000          (5)
2   0.083333          (9)
3   0.083333         (10)
4   0.083333         (13)
5   0.083333         (19)
6   0.083333         (24)
7   0.166667         (25)
8   0.125000         (33)
9   0.208333         (38)
10  0.250000         (40)
11  0.166667         (41)
12  0.083333      (4, 38)
13  0.083333      (40, 5)
14  0.083333      (41, 5)
15  0.083333     (41, 38)
16  0.083333     (40, 41)
17  0.083333  (40, 41, 5)

In [8]:
#Picking frequent itemsets now with names of items
apriori(df, min_support=0.05, use_colnames=True)

support                         itemsets
0   0.083333                  (bottled water)
1   0.125000                         (butter)
2   0.083333                      (chocolate)
3   0.083333                   (citrus fruit)
4   0.083333                   (cream cheese)
5   0.083333          (fruit/vegetable juice)
6   0.083333                     (newspapers)
7   0.166667               (other vegetables)
8   0.125000                     (rolls/buns)
9   0.208333                 (tropical fruit)
10  0.250000                     (whole milk)
11  0.166667                         (yogurt)
12  0.083333  (tropical fruit, bottled water)
13  0.083333             (butter, whole milk)
14  0.083333                 (yogurt, butter)
15  0.083333         (tropical fruit, yogurt)
16  0.083333             (yogurt, whole milk)
17  0.083333     (butter, yogurt, whole milk)

Next, we create the frequent itemsets via apriori and add a new column that stores the length of each itemset

In [9]:
frequent_itemsets = apriori(df, min_support=0.05, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets

support                         itemsets  length
0   0.083333                  (bottled water)       1
1   0.125000                         (butter)       1
2   0.083333                      (chocolate)       1
3   0.083333                   (citrus fruit)       1
4   0.083333                   (cream cheese)       1
5   0.083333          (fruit/vegetable juice)       1
6   0.083333                     (newspapers)       1
7   0.166667               (other vegetables)       1
8   0.125000                     (rolls/buns)       1
9   0.208333                 (tropical fruit)       1
10  0.250000                     (whole milk)       1
11  0.166667                         (yogurt)       1
12  0.083333  (tropical fruit, bottled water)       2
13  0.083333             (butter, whole milk)       2
14  0.083333                 (yogurt, butter)       2
15  0.083333         (tropical fruit, yogurt)       2
16  0.083333             (yogurt, whole milk)       2
17  0.083333     (butter, yogurt, whole milk)       3

In [10]:
#Exploring dataset according to specific criterias
frequent_itemsets[(frequent_itemsets['length'] == 3) &
                   (frequent_itemsets['support'] > 0.05)]

support                      itemsets  length
17  0.083333  (butter, yogurt, whole milk)       3

Here, we see that items that fulfills our criteria of having length being 3 and support being greater than the minimum spport is butter, whole milk and yogurt.

In [11]:
#Exploring datasets by specifying the items and finding its length and support
frequent_itemsets[frequent_itemsets['itemsets'] == {'yogurt', 'whole milk'} ]

support              itemsets  length
16  0.083333  (yogurt, whole milk)       2

Here, we see that for the item yogurt and butter, the length is 2 and the support is greater than the minimum support, i.e., 0.05.

**MINING ASSOCIATION RULE

We know that the association rules are simply the if-else statements. The IF component of an association rule is known as the antecedent. The THEN component is known as the consequent. The antecedent and the consequent are disjoint; they have no items in common.
So, here we create antecedents and consequents.

In [13]:
#We set our metric as "Lift" to define whether antecedents & consequents are dependent our not
rules = association_rules(frequent_itemsets , metric="lift", min_threshold=0.05)
rules

antecedents           consequents  antecedent support  \
0       (tropical fruit)       (bottled water)            0.208333   
1        (bottled water)      (tropical fruit)            0.083333   
2               (butter)          (whole milk)            0.125000   
3           (whole milk)              (butter)            0.250000   
4               (yogurt)              (butter)            0.166667   
5               (butter)              (yogurt)            0.125000   
6       (tropical fruit)              (yogurt)            0.208333   
7               (yogurt)      (tropical fruit)            0.166667   
8               (yogurt)          (whole milk)            0.166667   
9           (whole milk)              (yogurt)            0.250000   
10      (yogurt, butter)          (whole milk)            0.083333   
11  (whole milk, butter)              (yogurt)            0.083333   
12  (yogurt, whole milk)              (butter)            0.083333   
13              (butter)  (yogurt, whole milk)            0.125000   
14              (yogurt)  (whole milk, butter)            0.166667   
15          (whole milk)      (yogurt, butter)            0.250000   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.083333  0.083333    0.400000  4.800000  0.065972    1.527778  
1             0.208333  0.083333    1.000000  4.800000  0.065972         inf  
2             0.250000  0.083333    0.666667  2.666667  0.052083    2.250000  
3             0.125000  0.083333    0.333333  2.666667  0.052083    1.312500  
4             0.125000  0.083333    0.500000  4.000000  0.062500    1.750000  
5             0.166667  0.083333    0.666667  4.000000  0.062500    2.500000  
6             0.166667  0.083333    0.400000  2.400000  0.048611    1.388889  
7             0.208333  0.083333    0.500000  2.400000  0.048611    1.583333  
8             0.250000  0.083333    0.500000  2.000000  0.041667    1.500000  
9             0.166667  0.083333    0.333333  2.000000  0.041667    1.250000  
10            0.250000  0.083333    1.000000  4.000000  0.062500         inf  
11            0.166667  0.083333    1.000000  6.000000  0.069444         inf  
12            0.125000  0.083333    1.000000  8.000000  0.072917         inf  
13            0.083333  0.083333    0.666667  8.000000  0.072917    2.750000  
14            0.083333  0.083333    0.500000  6.000000  0.069444    1.833333  
15            0.083333  0.083333    0.333333  4.000000  0.062500    1.375000

The output above shows the values of various supporting components. To get more insights from the data, we sort the data by the confidence value.

In [15]:
#Sorting values based on confidence
rules.sort_values("confidence",ascending=True)

antecedents           consequents  antecedent support  \
3           (whole milk)              (butter)            0.250000   
9           (whole milk)              (yogurt)            0.250000   
15          (whole milk)      (yogurt, butter)            0.250000   
0       (tropical fruit)       (bottled water)            0.208333   
6       (tropical fruit)              (yogurt)            0.208333   
4               (yogurt)              (butter)            0.166667   
7               (yogurt)      (tropical fruit)            0.166667   
8               (yogurt)          (whole milk)            0.166667   
14              (yogurt)  (whole milk, butter)            0.166667   
2               (butter)          (whole milk)            0.125000   
5               (butter)              (yogurt)            0.125000   
13              (butter)  (yogurt, whole milk)            0.125000   
1        (bottled water)      (tropical fruit)            0.083333   
10      (yogurt, butter)          (whole milk)            0.083333   
11  (whole milk, butter)              (yogurt)            0.083333   
12  (yogurt, whole milk)              (butter)            0.083333   

    consequent support   support  confidence      lift  leverage  conviction  
3             0.125000  0.083333    0.333333  2.666667  0.052083    1.312500  
9             0.166667  0.083333    0.333333  2.000000  0.041667    1.250000  
15            0.083333  0.083333    0.333333  4.000000  0.062500    1.375000  
0             0.083333  0.083333    0.400000  4.800000  0.065972    1.527778  
6             0.166667  0.083333    0.400000  2.400000  0.048611    1.388889  
4             0.125000  0.083333    0.500000  4.000000  0.062500    1.750000  
7             0.208333  0.083333    0.500000  2.400000  0.048611    1.583333  
8             0.250000  0.083333    0.500000  2.000000  0.041667    1.500000  
14            0.083333  0.083333    0.500000  6.000000  0.069444    1.833333  
2             0.250000  0.083333    0.666667  2.666667  0.052083    2.250000  
5             0.166667  0.083333    0.666667  4.000000  0.062500    2.500000  
13            0.083333  0.083333    0.666667  8.000000  0.072917    2.750000  
1             0.208333  0.083333    1.000000  4.800000  0.065972         inf  
10            0.250000  0.083333    1.000000  4.000000  0.062500         inf  
11            0.166667  0.083333    1.000000  6.000000  0.069444         inf  
12            0.125000  0.083333    1.000000  8.000000  0.072917         inf

The above table shows the relationship between different items and the likelihood of a customer buying those items together. For example, according to the table, the customers who purchased whole milk are expected to buy butter with a likelihood of 33% (confidence).